### PR TITLE
refactor: gateway

### PR DESCRIPTION
**Pull request by atzoum**
# Description

Doing some preliminary work for being able to perform another refactoring that is going to modify gateway's functions to start relying on sourceID rather than writeKey:

- Stop using global (config) variables in the gateway package
- Stop using global config, logger & stats variables in `gateway.Handle`
- Organise `gateway.Handle` fields into logical groups
- Adapt `gateway#Setup` to fill in necessary configuration moved from global configuration
- Remove methods/variables that are no longer necessary
- Split `gateway.go` into more files containing logically grouped gateway functionalities
- Wait for gateway to initialise its backend configuration in gateway's tests

## Linear Ticket

[Link](https://linear.app/rudderstack/issue/INT-265/refactor-gateway-package)

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: gateway/gateway.go [hunk 1]
```diff
@@ -61,1612 +12,24 @@ import (
...
-	WriteTimeout                                                                      time.Duration
-	IdleTimeout                                                                       time.Duration
-	allowReqsWithoutUserIDAndAnonymousID                                              bool
-	gwAllowPartialWriteWithErrors                                                     bool
-	allowBatchSplitting                                                               bool
-	pkgLogger                                                                         logger.Logger
-	Diagnostics                                                                       diagnostics.DiagnosticsI
-	semverRegexp                                                                      = regexp.MustCompile(`^v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?$`)
+const (
+	delimiter                 = string("<<>>")
```
Comment (cisse21) [inline gateway/gateway.go:16]:
```suggestion
	delimiter                 = "<<>>"
```

File: gateway/handle_config.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	_, ok := gateway.conf.writeKeysSourceMap[writeKey]
+	return ok
+}
+
+func (gateway *Handle) isWriteKeyEnabled(writeKey string) bool {
+	gateway.conf.configSubscriberLock.RLock()
+	defer gateway.conf.configSubscriberLock.RUnlock()
+
+	return gateway.conf.writeKeysSourceMap[writeKey].Enabled
```
Comment (cisse21) [inline gateway/handle_config.go:46]: Minor

```suggestion
	source, ok := gateway.conf.writeKeysSourceMap[writeKey]; if !ok {
		return false
	}
	return source.Enabled
```

**Review by cisse21 — APPROVED** [review]

File: gateway/types.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	data     []byte
+	writeKey string
+}
+
+// Basic worker unit that works on incoming webRequests.
+//
+// Has three channels used to communicate between the two goroutines each worker runs.
+//
+// One to receive new webRequests, one to send batches of said webRequests and the third to receive errors if any in response to sending the said batches to dbWriterWorker.
+type userWebRequestWorkerT struct {
```
Comment (lvrach) [inline gateway/types.go:55]: [nit-pick]
```suggestion
// userWebRequestWorkerT is a basic worker unit that works on incoming webRequests.
//
// Has three channels used to communicate between the two goroutines each worker runs.
//
// One to receive new webRequests, one to send batches of said webRequests and the third to receive errors if any in response to sending the said batches to dbWriterWorker.
type userWebRequestWorkerT struct {
```

File: gateway/handle_http.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func (gateway *Handle) pixelWebRequestHandler(rh RequestHandler, w http.ResponseWriter, r *http.Request, reqType string) {
+	gateway.sendPixelResponse(w)
+	gateway.logger.LogRequest(r)
+	atomic.AddUint64(&gateway.recvCount, 1)
+	var errorMessage string
+	defer func() {
+		if errorMessage != "" {
+			gateway.logger.Info(fmt.Sprintf("IP: %s -- %s -- Error while handling request: %s", misc.GetIPFromReq(r), r.URL.Path, errorMessage))
```
Comment (lvrach) [inline gateway/handle_http.go:63]:
```suggestion
			gateway.logger.Infof("IP: %s -- %s -- Error while handling request: %s", misc.GetIPFromReq(r), r.URL.Path, errorMessage)
```

File: gateway/handle.go [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+
+		recvCount := atomic.LoadUint64(&gateway.recvCount)
+		ackCount := atomic.LoadUint64(&gateway.ackCount)
+		gateway.logger.Debug("Gateway Recv/Ack ", recvCount, ackCount)
+	}
+}
+
+// getPayloadFromRequest reads the request body and returns the payload's bytes or an error if the payload cannot be read
+// ProcessWebRequest is an Interface wrapper for webhook
+func (gateway *Handle) ProcessWebRequest(w *http.ResponseWriter, r *http.Request, reqType string, payload []byte, writeKey string) string {
```
Comment (lvrach) [inline gateway/handle.go:587]: This comment is incorrect

File: gateway/import_handler.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"net/http"
+	"strings"
+)
+
+// ImportRequestHandler is an empty struct to capture import specific request handling functionality
+type ImportRequestHandler struct{}
+
+// ProcessRequest on ImportRequestHandler splits payload by user and throws them into the webrequestQ and waits for all their responses before returning
+func (*ImportRequestHandler) ProcessRequest(gateway *Handle, w *http.ResponseWriter, r *http.Request, _ string, payload []byte, writeKey string) string {
```
Comment (lvrach) [inline gateway/import_handler.go:12]: Would doing something like this make sense? Is it planned for the next iteration ?

```suggestion
type ImportRequestHandler struct{
    gateway *Handle, 
}

// ProcessRequest on ImportRequestHandler splits payload by user and throws them into the webrequestQ and waits for all their responses before returning
func (*ImportRequestHandler) ProcessRequest(w *http.ResponseWriter, r *http.Request, _ string, payload []byte, writeKey string) string {
```

**Comment by atzoum** [comment]
> Most of the comments are irrelevant for this PR since large portions of code moved and this is preparatory PR.

tried to address them nevertheless :)

**Pull request merged**